versioned/storage/cache/src/main/java/org/projectnessie/versioned/storage/cache/CaffeineCacheBackend.java

```diff
@@ -100,34 +115,58 @@ public void clear(@Nonnull String repositoryId) {
     cache().asMap().keySet().removeIf(k -> k.repositoryId.equals(repositoryId));
   }
 
-  private CacheKey cacheKey(String repositoryId, ObjId id) {
-    return new CacheKey(repositoryId, id);
+  static CacheKeyValue cacheKey(String repositoryId, ObjId id) {
+    return new CacheKeyValue(repositoryId, id);
+  }
+
+  private static CacheKeyValue cacheKeyValue(
+      String repositoryId, ObjId id, byte[] value, long expiresAt) {
+    return new CacheKeyValue(repositoryId, id, value, expiresAt);
   }
 
-  static final class CacheKey {
+  /**
+   * Class used for both the cache key and cache value including the expiration timestamp. This is
+   * (should be) more efficient (think: mono-morphic vs bi-morphic call sizes) and more GC/heap
```

Comment:
Why is it "premature"? You'd need one object to have a composite key of repo-ID + obj-ID plus a holder for the value having the expiration and the actual value.

Reply:
Because in terms of space, having 2 separate objects, `CacheKey` and `CacheValue`, would take only 16 bytes more (2 object headers instead of one) when storing an entry, so it doesn't seem it would make a huge difference. OTOH, cache lookups only need the key, so having smaller keys would save some space.

As for GC... I don't know enough about GC internals to tell if it's more efficient to GC a big object, or 2 smaller ones. I've read that it's generally better to have fewer but larger objects indeed, but I've also read that G1, for instance, is sensitive to very large objects.

Anyways, it's OK to leave as is.

Reply:
Yes, G1 is very allergic to very large object, but "very large" means really many megabytes (humongous objects).